### PR TITLE
Feature/api

### PR DESCRIPTION
**Pull request by Galbator1x**

### CODE REVIEW

File: app/controllers/api/v1/tasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    q_params = params[:q] || { s: 'id asc' }
+
+    tasks = Task.all
+      .ransack(q_params)
+      .result
+      .page(params[:page])
+      .per(params[:per_page])
+
+    json = {
+      items: tasks.map { |t| TaskSerializer.new(t).as_json },
```
Comment (devality) [inline app/controllers/api/v1/tasks_controller.rb]: Look how we do it in PS project.

File: app/controllers/api/v1/tasks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  end
+
+  def create
+    task = current_user.my_tasks.new(task_params)
+
+    if task.save
+      respond_with task, location: nil
+    else
+      render json: { errors: task.errors }, status: :unprocessable_entity
+    end
```
Comment (devality) [inline app/controllers/api/v1/tasks_controller.rb]: And what about `respond_with`?

File: app/controllers/api/v1/tasks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
+class Api::V1::TasksController < Api::V1::ApplicationController
+  def index
+    q_params = params[:q] || { s: 'id asc' }
+
+    tasks = Task.all
```
Comment (pn-y) [inline app/controllers/api/v1/tasks_controller.rb]: s/Task.all/Task

File: app/controllers/api/v1/tasks_controller.rb [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+      respond_with task, location: nil
+    else
+      respond_with json: { errors: task.errors }, status: :unprocessable_entity
+    end
+  end
+
+  def update
+    task = Task.find(params[:id])
+
+    if task.update(task_params)
```
Comment (pn-y) [inline app/controllers/api/v1/tasks_controller.rb]: Doesn't that covered by responders?
```
task = Task.find(params[:id])
task.update(task_params)
respond_with task (and maybe location: nil)

File: app/serializers/task_serializer.rb [hunk 5]
```diff
@@ -0,0 +1,9 @@
+class TaskSerializer < ActiveModel::Serializer
+  attributes :id, :name, :description, :state, :expired_at
+  belongs_to :author
+  belongs_to :assignee
+
+  def id
+    object.id.to_s
```
Comment (pn-y) [inline app/serializers/task_serializer.rb]: Why do you need string id?

File: app/controllers/api/v1/tasks_controller.rb [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+    q_params = params[:q] || { s: 'id asc' }
+
+    tasks = Task.all
+      .ransack(q_params)
+      .result
+      .page(params[:page])
+      .per(params[:per_page])
+
+    json = {
+      items: serialize(tasks, TaskSerializer),
```
Comment (MrPhantomT) [inline app/controllers/api/v1/tasks_controller.rb]: ActiveModelSerializer can serialize you collections by himself you just need to pass serializer to `serializer` or `each_serializer` option in the render method

File: app/controllers/api/v1/tasks_controller.rb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+      meta: build_meta(tasks)
+    }
+
+    respond_with json
+  end
+
+  def show
+    task = Task.find(params[:id])
+    respond_with task
```
Comment (MrPhantomT) [inline app/controllers/api/v1/tasks_controller.rb]: Why you don't serialize task?

Comment (MrPhantomT) [reply app/controllers/api/v1/tasks_controller.rb]: You need to serialize each object which gonna be returned from you API

File: test/factories/tasks.rb [hunk 8]
```diff
@@ -1,9 +1,8 @@
 FactoryBot.define do
   factory :task do
-    author
-    assignee
     name { generate :string }
     description { generate :string }
-    expired_at { 1.month.from_now }
+    author_id { create :user }
```
Comment (MrPhantomT) [inline test/factories/tasks.rb]: Strange, You have tried to initialize author_id with an instance of a user but not with his id.
You could define has_one associations in factory bot like this:
```
association :author, factory: :user
```

**Comment by MrPhantomT** [comment]
👍

**Pull request merged**